Duplication of libLLVM-*.so in the distributed rustc component.

Since #57286, `libLLVM-*.so` has been distributed in the `rustc` component in two locations:
* `lib/libLLVM-*.so` (presumably as a runtime dependency of `bin/rustc`)
* `lib/rustlib/$target/lib/libLLVM-*.so` (not sure, maybe for *linking* custom drivers?)
  * likely belongs in the `rustc-dev` component, *if* the duplication can't be avoided

<hr/>

The difference can be seen in these consecutive nightlies' `rustc` components:
* https://static.rust-lang.org/dist/2019-01-06/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
  * has just *the second* `libLLVM-*.so`
    * maybe it's used via relative `RPATH` by `lib/rustlib/$target/codegen-backends/librustc_codegen_llvm-llvm.so`? separate codegen backends are gone, maybe we should just try removing it?
* https://static.rust-lang.org/dist/2019-01-07/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
  * first nightly to include #57286
  * has both copies

All 3 of those `libLLVM-8svn.so` files are identical (even across nightlies, yay for deterministic builds):
```
e400d157a7ceaee26f8680cb6f7430bd98c251dbb32e1f10549dda9f5769dc42  before/rustc-nightly-x86_64-unknown-linux-gnu/rustc/lib/rustlib/x86_64-unknown-linux-gnu/lib/libLLVM-8svn.so
e400d157a7ceaee26f8680cb6f7430bd98c251dbb32e1f10549dda9f5769dc42  after/rustc-nightly-x86_64-unknown-linux-gnu/rustc/lib/libLLVM-8svn.so
e400d157a7ceaee26f8680cb6f7430bd98c251dbb32e1f10549dda9f5769dc42  after/rustc-nightly-x86_64-unknown-linux-gnu/rustc/lib/rustlib/x86_64-unknown-linux-gnu/lib/libLLVM-8svn.so
```

One fascinating aspect is that the size only increases from `82.22 MiB` to `83.18 MiB`, despite `libLLVM-8svn.so` alone being `67 MiB` uncompressed, so compression is really helping.

Even the `.tar.gz` only goes from `111.41 MiB` to `113.26 MiB`.
(I'm using archive sizes from the [2019-01-06](https://static.rust-lang.org/dist/2019-01-06) and [2019-01-07](https://static.rust-lang.org/dist/2019-01-07) directory listings in https://static.rust-lang.org/dist, FWIW)

<hr/>

This part of the build system seems relevant, but I'm not sure which of the two this creates: https://github.com/rust-lang/rust/blob/733f104f138477ef20f59eadcbcf8800b21a7732/src/bootstrap/compile.rs#L755-L757

cc @alexcrichton @rust-lang/release

Hmm, I can't actually get cross-builds to work with `rustc-dev`. I added this to `rustc.rs`:

```rust
#![feature(rustc_private)]
extern crate rustc_driver;
```

... so it would pick that up from the sysroot. Trying to cross-compile i686, I get:

```
error[E0463]: can't find crate for `rustc_macros` which `rustc_driver` depends on
  --> ./rustc.rs:30:5
   |
30 |     rustc_driver::set_sigpipe_handler();
   |     ^^^^^^^^^^^^ can't find crate
```

I do have both i686 and x86_64 builds of `rustc_macros` in the sysroot. I suspect the error is because the i686 build has different metadata than my host build, so it can't find the one it wants for the host. That may mean that `rustc-dev` is really tied to being used on the same host.

It works if I just compile native x86_64 though. Then I tried removing the `rustlib/.../libLLVM.so`, and it gets essentially the same error as #68714:

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" [...] "-lLLVM-9-rust-1.44.0-nightly" "-lutil" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: /usr/bin/ld: cannot find -lLLVM-9-rust-1.44.0-nightly
          collect2: error: ld returned 1 exit status
```

AFAICS, building normal crates still works fine with that extra `libLLVM` removed though.

----

In short:
- Cross-compiling with `rustc-dev` seems broken -- maybe it should be shipped native-only.
  - The only difference would be in the manifest created for rustup.
- Linking to `rustc-dev` crates will inherit `-lLLVM`, so that needs to be in the library path.

--- a/src/bootstrap/compile.rs
+++ b/src/bootstrap/compile.rs
@@ -772,7 +772,8 @@ impl Step for Assemble {
 
         // Ensure that `libLLVM.so` ends up in the newly build compiler directory,
         // so that it can be found when the newly built `rustc` is run.
-        dist::maybe_install_llvm_dylib(builder, target_compiler.host, &sysroot);
+        dist::maybe_install_llvm_runtime(builder, target_compiler.host, &sysroot);
+        dist::maybe_install_llvm_target(builder, target_compiler.host, &sysroot);
 
         // Link the compiler binary itself into place
         let out_dir = builder.cargo_out(build_compiler, Mode::Rustc, host);

--- a/src/bootstrap/dist.rs
+++ b/src/bootstrap/dist.rs
@@ -514,7 +514,7 @@ impl Step for Rustc {
             // components like the llvm tools and LLD. LLD is included below and
             // tools/LLDB come later, so let's just throw it in the rustc
             // component for now.
-            maybe_install_llvm_dylib(builder, host, image);
+            maybe_install_llvm_runtime(builder, host, image);
 
             // Copy over lld if it's there
             if builder.config.lld_enabled {
@@ -2228,27 +2228,18 @@ impl Step for HashSign {
     }
 }
 
-// Maybe add libLLVM.so to the lib-dir. It will only have been built if
-// LLVM tools are linked dynamically.
-//
-// We add this to both the libdir of the rustc binary itself (for it to load at
-// runtime) and also to the target directory so it can find it at link-time.
-//
-// Note: This function does no yet support Windows but we also don't support
-//       linking LLVM tools dynamically on Windows yet.
-pub fn maybe_install_llvm_dylib(builder: &Builder<'_>, target: Interned<String>, sysroot: &Path) {
+/// Maybe add libLLVM.so to the given destination lib-dir. It will only have
+/// been built if LLVM tools are linked dynamically.
+///
+/// Note: This function does not yet support Windows, but we also don't support
+///       linking LLVM tools dynamically on Windows yet.
+fn maybe_install_llvm(builder: &Builder<'_>, target: Interned<String>, dst_libdir: &Path) {
     let src_libdir = builder.llvm_out(target).join("lib");
-    let dst_libdir1 = sysroot.join("lib/rustlib").join(&*target).join("lib");
-    let dst_libdir2 =
-        sysroot.join(builder.sysroot_libdir_relative(Compiler { stage: 1, host: target }));
-    t!(fs::create_dir_all(&dst_libdir1));
-    t!(fs::create_dir_all(&dst_libdir2));
 
     if target.contains("apple-darwin") {
         let llvm_dylib_path = src_libdir.join("libLLVM.dylib");
         if llvm_dylib_path.exists() {
-            builder.install(&llvm_dylib_path, &dst_libdir1, 0o644);
-            builder.install(&llvm_dylib_path, &dst_libdir2, 0o644);
+            builder.install(&llvm_dylib_path, dst_libdir, 0o644);
         }
         return;
     }
@@ -2262,9 +2253,21 @@ pub fn maybe_install_llvm_dylib(builder: &Builder<'_>, target: Interned<String>,
             panic!("dist: Error calling canonicalize path `{}`: {}", llvm_dylib_path.display(), e);
         });
 
-        builder.install(&llvm_dylib_path, &dst_libdir1, 0o644);
-        builder.install(&llvm_dylib_path, &dst_libdir2, 0o644);
+        builder.install(&llvm_dylib_path, dst_libdir, 0o644);
     }
+}
+
+/// Maybe add libLLVM.so to the target lib-dir for linking.
+pub fn maybe_install_llvm_target(builder: &Builder<'_>, target: Interned<String>, sysroot: &Path) {
+    let dst_libdir = sysroot.join("lib/rustlib").join(&*target).join("lib");
+    maybe_install_llvm(builder, target, &dst_libdir);
+}
+
+/// Maybe add libLLVM.so to the runtime lib-dir for rustc itself.
+pub fn maybe_install_llvm_runtime(builder: &Builder<'_>, target: Interned<String>, sysroot: &Path) {
+    let dst_libdir =
+        sysroot.join(builder.sysroot_libdir_relative(Compiler { stage: 1, host: target }));
+    maybe_install_llvm(builder, target, &dst_libdir);
 }
 
 #[derive(Clone, Debug, Eq, Hash, PartialEq)]
@@ -2313,6 +2316,12 @@ impl Step for LlvmTools {
             let exe = src_bindir.join(exe(tool, &target));
             builder.install(&exe, &dst_bindir, 0o755);
         }
+
+        // Copy libLLVM.so to the target lib dir as well, so the RPATH like
+        // `$ORIGIN/../lib` can find it. It may also be used as a dependency
+        // of `rustc-dev` to support the inherited `-lLLVM` when using the
+        // compiler libraries.
+        maybe_install_llvm_target(builder, target, &image);
 
         // Prepare the overlay
         let overlay = tmp.join("llvm-tools-overlay");

